nnstreamer/tensor_trainer/tensor_trainer_nntrainer.cc

```diff
@@ -39,19 +42,107 @@ void init_subplugin_nntrainer(void) __attribute__((constructor));
  */
 void fini_subplugin_nntrainer(void) __attribute__((destructor));
 
+static void nntrainer_model_destructor(const GstTensorFilterProperties *prop,
+                                       void **private_data) {
+  NNTrainer::NNTrainerTrain *nntrainer =
+    static_cast<NNTrainer::NNTrainerTrain *>(*private_data);
+
+  if (!nntrainer)
+    return;
+  delete nntrainer;
+  *private_data = NULL;
+}
+
+#if 0 /** To avoid build errors : defined but not used */
+static void nntrainer_model_train(const GstTensorFilterProperties *prop,
+                                  void **private_data) {
+  NNTrainer::NNTrainerTrain *nntrainer =
+    static_cast<NNTrainer::NNTrainerTrain *>(*private_data);
+  if (!nntrainer) {
+    std::cerr << "Failed get model" << std::endl;
+  }
+
+  try {
+    nntrainer->trainModel();
+  } catch (std::exception &e) {
+    std::cerr << "Error" << typeid(e).name() << "," << e.what() << std::endl;
+  }
+}
+#endif
+
+void NNTrainer::NNTrainerTrain::trainModel() {
+  try {
+    model->train();
+    training_loss = model->getTrainingLoss();
+    validation_loss = model->getValidationLoss();
+  } catch (std::exception &e) {
+    std::cerr << "Error" << typeid(e).name() << "," << e.what() << std::endl;
+  }
+}
+
+void NNTrainer::NNTrainerTrain::createModel() {
+
+  try {
+    model = ml::train::createModel(ml::train::ModelType::NEURAL_NET);
+    model->load(model_config,
+                ml::train::ModelFormat::MODEL_FORMAT_INI_WITH_BIN);
+  } catch (std::exception &e) {
+    std::cerr << "Failed loading model configuration file" << typeid(e).name()
+              << e.what() << std::endl;
+  }
+
+  try {
+    model->compile();
+    model->initialize();
+    model->setDataset(ml::train::DatasetModeType::MODE_TRAIN, dataset_train);
+    model->setDataset(ml::train::DatasetModeType::MODE_VALID, dataset_valid);
+  } catch (std::exception &e) {
+    std::cerr << "Error" << typeid(e).name() << "," << e.what() << std::endl;
+  }
+}
+
+NNTrainer::NNTrainerTrain::NNTrainerTrain(const std::string &_model_config) :
+  model_config(_model_config) {
+  createModel();
+}
+
+static int
+nntrainer_model_construct_with_conf(const GstTensorFilterProperties *prop,
+                                    void **private_data) {
+
+  NNTrainer::NNTrainerTrain *nntrainer =
+    static_cast<NNTrainer::NNTrainerTrain *>(*private_data);
+  if (!nntrainer)
+    nntrainer_model_destructor(prop, private_data);
+
+  try {
+    nntrainer =
```

Comment:
nntrainer is already initialized at line 113 so I wonder why does need to make a new object?